### PR TITLE
feat: normalize patterns to handle "./" prefix in files and ignores

### PR DESCRIPTION
**Pull request by Pixel998**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request?

Addresses issue where patterns starting with "./" were not correctly matched in `files` and `ignores`.

#### What changes did you make? (Give an overview)

Strip "./" from the start of strings in `files` and `ignores` so that they match as expected.

#### Related Issues

https://github.com/eslint/eslint/issues/18757

### CODE REVIEW

**Review by nzakas — COMMENTED** [review]
Thanks for taking a look at this. Overall, it looks good.

I think we have a question here: Is it safe to overwrite `files` in an existing object or should we be creating a new object? I think we may need to create a new object to be safe, but that would also increase the number of objects we create. And if we do that, perhaps we should only do so when `files` has at least one pattern starting with `"./"` and reuse every other config as-is?

@eslint/eslint-tsc what do you think?

**Comment by mdjermanovic** [comment]
I think we should create a new object when we need to modify its content. In this case, where at least one of its `files` / `ignores` patterns starts with `"./"` or `"!./"`.

File: packages/config-array/tests/config-array.test.js [hunk 1]
```diff
@@ -1229,6 +1229,149 @@ describe("ConfigArray", () => {
...
+					);
+
+					configs.normalizeSync();
+
+					assert.strictEqual(
+						configs.getConfig("src/foo.js").defs.severity,
+						"error",
+					);
+				});
+			});
```
Comment (mdjermanovic) [inline packages/config-array/tests/config-array.test.js:1374]: Can you add a test with `configs.normalize()` to ensure patterns will be normalized when it's used? (currently, all tests are with `configs.normalizeSync()`)

File: packages/config-array/src/config-array.js [hunk 2]
```diff
@@ -244,6 +244,44 @@ function doMatch(filepath, pattern, options = {}) {
...
+/**
+ * Normalizes `files` and `ignores` patterns in a config by removing "./" prefixes.
+ * @param {Object} config The config object to normalize patterns in.
+ */
+function normalizeConfigPatterns(config) {
+	if (Array.isArray(config.files)) {
+		config.files = config.files.map(pattern => {
+			if (Array.isArray(pattern)) {
+				return pattern.map(normalizePattern);
+			}
```
Comment (mdjermanovic) [inline packages/config-array/src/config-array.js:275]: Can you also add a test for this case?

File: packages/config-array/tests/config-array.test.js
Comment (nzakas) [inline packages/config-array/tests/config-array.test.js:1]: Can you add a test that verifies a different object is created when replacing `./` and `!./`?

File: packages/config-array/src/config-array.js [hunk 3]
```diff
@@ -244,6 +244,80 @@ function doMatch(filepath, pattern, options = {}) {
...
+		isString(pattern) &&
+		(pattern.startsWith("./") || pattern.startsWith("!./"))
+	);
+}
+
+/**
+ * Normalizes `files` and `ignores` patterns in a config by removing "./" prefixes.
+ * @param {Object} config The config object to normalize patterns in.
+ */
+function normalizeConfigPatterns(config) {
```
Comment (mdjermanovic) [inline packages/config-array/src/config-array.js:283]:
```suggestion
function normalizeConfigPatterns(config) {
	if (!config) {
		return config;
	}
```

I've just tried this version of config-array with eslint/eslint, and several tests that check how nullish config elements are reported are failing with errors like `Cannot read properties of undefined (reading 'files')`. I think we can address this by adding an early exit for that case.

**Comment by mdjermanovic** [comment]
I changed the tag to `feat` as this enables patterns that are currently no-op.

File: packages/config-array/src/config-array.js [hunk 4]
```diff
@@ -244,6 +244,84 @@ function doMatch(filepath, pattern, options = {}) {
...
+	return (
+		isString(pattern) &&
+		(pattern.startsWith("./") || pattern.startsWith("!./"))
+	);
+}
+
+/**
+ * Normalizes `files` and `ignores` patterns in a config by removing "./" prefixes.
+ * @param {Object} config The config object to normalize patterns in.
+ */
```
Comment (nzakas) [inline packages/config-array/src/config-array.js:282]:
```suggestion
 * @returns {Object} The normalized config object.
 */
```

File: packages/config-array/src/config-array.js (hunk 3 above)
Comment (nzakas) [reply packages/config-array/src/config-array.js:283]: Can you also add a test with `null` and `undefined` configs in this PR?

**Review by nzakas — APPROVED** [review]
LGTM, thanks!

Would like @mdjermanovic to approve before merging.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!
